src/workerd/io/observer.h

```diff
@@ -30,6 +30,18 @@ class WebSocketObserver: public kj::Refcounted {
   virtual void receivedMessage(size_t bytes) {};
 };
 
+// Observes a byte stream. This interface should only be used by byte streams which internally queue
+// each write and each write of size N is accompanied by a corresponding read of the same size.
+// Thus, the current number of items in the internal queue can be calculated as the number of times
+// onWrite() has been called minus the number of times onRead() has been called.
+class ByteStreamObserver {
+public:
+  // Called when a write of size `bytes` is enqueued on the stream.
+  virtual void onWrite(size_t bytes) {};
+  // Called when a write of size `bytes` is dequeued from the stream.
```

Comment:
```suggestion
  // Called when a read of size `bytes` is dequeued from the stream.
```

Reply:
I think I might be better off renaming `write` to `item` and `onWrite` to `enqueued` and `dequeued` to better reflect that we are observing operations on the internal queueing supported by streams rather than, for example, writing to the stream sink. WDYT?

Reply:
I don't know, I feel like that's less descriptive.
maybe `onWriteEvent` `onWriteEnqueued` `onWriteEventEnqueued` ?

Reply:
The term `chunk` seems to be used in the code and in the [writable stream JS API docs](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) so I'll go for `onChunkEnqueued` and `onChunkDequeued`.

Reply:
SGTM